### PR TITLE
RES933-Cifar100

### PR DESCRIPTION
**Pull request by lucasosouza**
Experiments with CIFAR-100
Can be extended to any torch vision dataset
[RES933](https://ntaresearch.atlassian.net/browse/RES-933)

### CODE REVIEW

**Comment by lucasosouza** [comment]
@subutai  please review

File: projects/whydense/cifar-100/cifar_experiment.py [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+                                         train=False,
+                                         transform=transform_noise_test)
+    loaders.append(
+      DataLoader(testset, batch_size=batch_size, shuffle=False)
+    )
+
+  return loaders
+
+
+class TinyCIFAR(object):
```
Comment (subutai) [inline projects/whydense/cifar-100/cifar_experiment.py]: We should change the name of the class to make it more unique.

File: projects/whydense/cifar-100/cifar_noise_test.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+  print("layer:", layer_name, "mean percent_on: ", percent_on / numImages)
+
+
+def testModelSparsity(config, options, projectDir):
+  """
+  Test a pretrained network, specified in this config against noisy vectors
+  """
+
+  # Pre-download dataset
+  data_dir = os.path.join(projectDir, "data")
```
Comment (subutai) [inline projects/whydense/cifar-100/cifar_noise_test.py]: CIFAR100 here right?

File: projects/whydense/cifar-100/cifar_noise_test.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+
+
+def layer_noise_sensitivities(config, options, projectDir):
+  """
+  Test a pretrained network, specified in this config against noisy vectors
+  """
+
+  # Pre-download dataset
+  data_dir = os.path.join(projectDir, "data")
+  _ = datasets.CIFAR10(data_dir, download=True, train=False)
```
Comment (subutai) [inline projects/whydense/cifar-100/cifar_noise_test.py]: CIFAR100 here right?

File: projects/whydense/cifar-100/vgg_exploration.cfg [hunk 4]
```diff
@@ -0,0 +1,984 @@
...
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+[DEFAULT]
+# changed path to a more generic location outside project folder
+# will allow comparison in tensorboard
+path = "~/nta/results"
```
Comment (subutai) [inline projects/whydense/cifar-100/vgg_exploration.cfg]: Are you using this file? If not, we can delete it.

File: projects/whydense/cifar-100/cifar_experiment.py [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+  """
+  Create a list of data loaders, one for each noise value
+  """
+  print("Creating test loaders for noise values:", noise_values)
+  loaders = []
+  for noise in noise_values:
+
+    transform_noise_test = transforms.Compose([
+      transforms.ToTensor(),
+      transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
```
Comment (subutai) [inline projects/whydense/cifar-100/cifar_experiment.py]: Did you verify this was the correct normalization for CIFAR-100? I took it from a CIFAR-10 sample, but not sure whether it should be the same for CIFAR100.

File: projects/whydense/cifar-100/cifar_experiment.py [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+      torch.cuda.manual_seed(seed)
+    else:
+      print("setup: Using cpu")
+      self.device = torch.device("cpu")
+
+    self.transform_train = transforms.Compose([
+      transforms.RandomCrop(32, padding=4),
+      transforms.RandomHorizontalFlip(),
+      transforms.ToTensor(),
+      transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
```
Comment (subutai) [inline projects/whydense/cifar-100/cifar_experiment.py]: Just realized - isn't it incorrect to normalize after cropping? Shouldn't people normalize before cropping?

File: projects/whydense/cifar-100/cifar_experiment.py [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+        self.learning_rate = np.mean(self.lr_scheduler.get_lr())
+
+  def _early_stopping(self, epoch, metric):
+    """ Custom early stopping based on moving median """ 
+
+    self.mean_losses.append(metric)
+    if metric >= np.median(self.mean_losses):
+      self.bad_epoches += 1
+    else:
+      self.bad_epoches = 0
```
Comment (subutai) [inline projects/whydense/cifar-100/cifar_experiment.py]: Should be "bad_epochs"

**Comment by lucasosouza** [comment]
Will merge now so Luiz can go ahead and merge the style change, and do the requested modifications right after on a new commit.
